examples/research_projects/codeparrot/scripts/pretokenizing.py

```diff
@@ -0,0 +1,49 @@
+import multiprocessing
+import time
+
+from datasets import load_dataset
+
+from arguments import PretokenizationArguments
+from transformers import AutoTokenizer, HfArgumentParser
+
+
+def tokenize(example):
+    output = dict()
+    output["input_ids"] = tokenizer(example["content"], truncation=False)["input_ids"]
+    output["ratio_char_token"] = len(example["content"]) / len(output["input_ids"])
+    return output
+
+
+parser = HfArgumentParser(PretokenizationArguments)
+args = parser.parse_args()
+if args.num_workers is None:
+    args.num_workers = multiprocessing.cpu_count()
+tokenizer = AutoTokenizer.from_pretrained(args.tokenizer_dir)
+
+t_start = time.time()
+ds = load_dataset(args.dataset_name, split="train")
+print(f"Dataset loaded in {time.time()-t_start:.2f}s")
+
+t_start = time.time()
+ds = ds.map(
+    tokenize,
+    num_proc=args.num_workers,
```

Comment:
I am not sure parallelizing much on the Python side helps here a lot. The tokenizer already tokenizes in parallel efficiently in Rust and so passing batches to the `tokenize` function is probably more efficient.
```suggestion
    batched=True,
    batch_size=10_000,
```
Note that you need to change your `tokenize` function slightly since it will get batches instead if single samples.

Reply:
I made a test on the validation set, it seems like relying on the tokenizer and batches slows things, it takes 73s to tokenize this way while it takes 38s to tokenize using python multiprocessing (nvidia a100)